src/node_wasm_streaming.cc

```diff
@@ -0,0 +1,377 @@
+    client_.reset();
+    compiled_module_bytes_.Reset();
+  }
+};
+
+WasmStreaming::WasmStreaming(Environment* env,
+                             Local<Object> wrap,
+                             std::shared_ptr<v8::WasmStreaming> wasm_streaming)
+  : BaseObject(env, wrap), wasm_streaming_(wasm_streaming) {
+  client_ = std::make_shared<Client>();
+  wasm_streaming_->SetClient(client_);
+  MakeWeak();
+}
+
+//  Pass a new chunk of bytes to WebAssembly streaming compilation.
+void WasmStreaming::OnBytesReceived(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+
+  WasmStreaming* ws;
+  ASSIGN_OR_RETURN_UNWRAP(&ws, args.Holder());
+
+  if (args.Length() == 0 || !args[0]->IsArrayBufferView()) {
+    env->ThrowError("First argument must be an ArrayBufferView");
+    return;
+  }
+
+  auto const* p = reinterpret_cast<uint8_t *>(Buffer::Data(args[0]));
+  if (ws->wasm_streaming_)
+    ws->wasm_streaming_->OnBytesReceived(p, Buffer::Length(args[0]));
+    // (makes a copy)
+}
+
+// Finish should be called after all received bytes were passed to
+// .onBytesReceived to tell V8 that there will be no more bytes.
+//
+// Optionally accepts a callback function to be invoked when compilation
+// completes.  The callback receive compiled module bytes.
+void WasmStreaming::Finish(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+
+  WasmStreaming* ws;
+  ASSIGN_OR_RETURN_UNWRAP(&ws, args.Holder());
+
+  if (args.Length() != 0) {
+    if (!args[0]->IsFunction()) {
+      env->ThrowError("First argument must be a function");
+      return;
+    }
+    if (ws->client_) {
+      auto reg = std::make_unique<OnModuleCompiledCallbackReg>(
+          env, args[0].As<Function>());
+      ws->client_->InstallCallback(std::move(reg));
+    }
+  }
+
+  if (ws->wasm_streaming_)
+    ws->wasm_streaming_->Finish();
+
+  ws->Reset();
+}
+
+// Abort streaming compilation with exception.
+void WasmStreaming::Abort(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+
+  WasmStreaming* ws;
+  ASSIGN_OR_RETURN_UNWRAP(&ws, args.Holder());
+
+  if (args.Length() == 0) {
+    env->ThrowError("No arguments provided");
+    return;
+  }
+
+  if (ws->wasm_streaming_)
+    ws->wasm_streaming_->Abort(args[0]);
+
+  ws->Reset();
+}
+
+// Passes previously compiled module bytes. This must be called before
+// .onBytesReceived, .finish, or .abort. Returns true if the module bytes
+// can be used, false otherwise.
+void WasmStreaming::SetCompiledModuleBytes(
+                                      const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  Isolate* isolate = env->isolate();
+
+  WasmStreaming* ws;
+  ASSIGN_OR_RETURN_UNWRAP(&ws, args.Holder());
+
+  if (args.Length() == 0 || !args[0]->IsArrayBufferView()) {
+    env->ThrowError("First argument must be an ArrayBufferView");
```

Comment:
At some point somebody is going to mention this, so: We typically use the errors from `node_errors.h` for things like this, there should be one for invalid types already specified in there